### PR TITLE
CPLAT-11143 Emit a build error for mismatched components in the new boilerplate

### PR DESCRIPTION
**Pull request by greglittlefield-wf**
## Motivation
If a component isn't named correctly in the new boilerplate, it won't get associated with the factory and props class the author intended it to be.

Currently, this mismatched component is ignored and the component gets interpreted as a props mapview, and when the consumer invokes the builder, they get a runtime error.

It would be better if it failed the build.

## Changes
- When possible, increase `v4_mixinBased` confidence of components to above `neutral`, resulting in mismatched components emitting errors instead of getting silently ignored.
   - Add tests
- Update runtime error to mention the mismatched component case, for better debugging of edge case mismatched components (with neutral confidences) that don't trigger a build error
- (Boyscouting) update punctuation in error messages

#### Release Notes
Emit a build error for mismatched components in the new boilerplate

## Review
_[See CONTRIBUTING.md][contributing-review-types] for more details on review types (+1 / QA +1 / +10) and code review process._

  

Please review: 

### QA Checklist
- [ ] Tests were updated and provide good coverage of the changeset and other affected code
- [ ] Manual testing was performed if needed
    - [ ] Steps from PR author: 
        - [ ] Verify in consumer repos that this does not cause any new failures
    - [ ] Anything falling under manual testing criteria [outlined in CONTRIBUTING.md][contributing-manual-testing]

## Merge Checklist
While we perform many automated checks before auto-merging, some manual checks are needed:
- [ ] A Client Platform member has reviewed these changes
- [ ] There are no unaddressed comments _- this check can be automated if reviewers use the "Request Changes" feature_
- [ ] _For release PRs -_ Version metadata in Rosie comment is correct

[contributing-review-types]: https://github.com/Workiva/over_react/blob/master/CONTRIBUTING.md#review-types
[contributing-manual-testing]: https://github.com/Workiva/over_react/blob/master/CONTRIBUTING.md#manual-testing-criteria

### CODE REVIEW

**Comment by aviary-wf** [comment]
## Security Insights
No security relevant content was detected by automated scans.

## Action Items
* Review PR for [security impact](https://wiki.atl.workiva.net/display/SECURITY/Development+Security+Review+Guidelines); comment "security review required" if needed or unsure
* Verify [`aviary.yaml`](https://dev.workiva.net/docs/teams/information-security/aviary#configuring-aviary) coverage of security relevant code
---
_Questions or Comments? Reach out on Slack: #support-infosec._

File: lib/src/builder/parsing/declarations_from_members.dart [hunk 1]
```diff
@@ -362,7 +362,7 @@ Iterable<BoilerplateDeclaration> getBoilerplateDeclarations(
 
 const errorStateOnly =
     'Could not find matching factory, props class, and component class in this file;'
-    ' these are required to use UiState';
+    ' these are required to use UiState.';
```
Comment (willdrach-wk) [inline lib/src/builder/parsing/declarations_from_members.dart:365]: Yay! Consistent grammar!

File: lib/src/builder/parsing/members_from_ast.dart [hunk 2]
```diff
@@ -429,26 +426,54 @@ class _BoilerplateMemberDetector {
...
+          //
+          // Make this `likely` so that components that don't get associated with factory/props
+          // due to naming issues aren't silently ignored.
+          v4_mixinBased: !classish.hasAbstractKeyword &&
+                  !_overridesIsClassGenerated(classish) &&
+                  mixinBoilerplateBaseClasses.contains(classish.superclass?.nameWithoutPrefix)
+              ? Confidence.likely
+              : Confidence.neutral,
+          v2_legacyBackwardsCompat: Confidence.neutral,
+          v3_legacyDart2Only: Confidence.neutral,
```
Comment (evanweible-wf) [inline lib/src/builder/parsing/members_from_ast.dart:463]: This is really cool

File: lib/src/builder/parsing/members_from_ast.dart [hunk 3]
```diff
@@ -429,26 +426,54 @@ class _BoilerplateMemberDetector {
     return false;
   }
 
+  /// UiComponent, UiComponent2, UiStatefulComponent, FluxUiComponent, CustomUiComponent, ...
+  static final _componentBaseClassPattern = RegExp('Ui\w+Component');
```
Comment (willdrach-wk) [inline lib/src/builder/parsing/members_from_ast.dart:430]: This doesn't actually match `FluxUiComponent` or `CustomUiComponent`, which might be fine but should update the doc comment.

```suggestion
  static final _componentBaseClassPattern = RegExp('.*Ui\w+Component');
```

Comment (greglittlefield-wf) [reply lib/src/builder/parsing/members_from_ast.dart:430]: Good catch!  I think it actually needs to be
```suggestion
  static final _componentBaseClassPattern = RegExp(r'Ui\w*Component');
```
though.

Also FYI the regex matches even if it's a partial match:
```dart
main() {
  final _componentBaseClassPattern = RegExp(r'Ui\w*Component');
  const names = [
    'UiComponent', 
    'UiComponent2', 
    'UiStatefulComponent',
    'FluxUiComponent', 
    'CustomUiComponent',
  ];
  
  for (var name in names) {
    final hasMatch = _componentBaseClassPattern.hasMatch(name);
    print('hasMatch for "$name": $hasMatch');
  }
}
```
```
hasMatch for "UiComponent": true
hasMatch for "UiComponent2": true
hasMatch for "UiStatefulComponent": true
hasMatch for "FluxUiComponent": true
hasMatch for "CustomUiComponent": true
```

File: lib/src/builder/parsing/members_from_ast.dart (hunk 2 above)
Comment (greglittlefield-wf) [reply lib/src/builder/parsing/members_from_ast.dart:463]: Thanks! I'm super happy that the confidence approach made this kind of change easy

**Comment by greglittlefield-wf** [comment]
- Regression tests were added for the issue
- The component that experienced this issue resulted in a build error
- Building wdesk 1.5.3499 w/ this version of over_react resulted in no changes in generated Dart files
- All classes outside of `lib/` (e.g., not verified by the wdesk check) that could be affected by this change are part of boilerplate
    https://sourcegraph.wk-dev.wdesk.org/search?q=lang:dart+%28%5Cn%7C%5E///.*%29%5Cn%5Eclass%5B%5E%5C%7B%3B%5D%2B%28%28%3F:Flux%29%3FUi%28%3F:Stateful%29%3FComponent%292+-repo:over_react+count:1000+file:%28app%7Cexample%7Ctest%7Ctool%29/&patternType=regexp


QA +1

**Comment by greglittlefield-wf** [comment]
@Workiva/release-management-p
